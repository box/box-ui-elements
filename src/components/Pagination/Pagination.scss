@import '../variables';

.be-pagination {
    display: flex;
}

.be-pagination-count {
    align-items: center;
    display: flex;
    margin-right: 20px;
}
