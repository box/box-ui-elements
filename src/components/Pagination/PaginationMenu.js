/**
 * @flow
 * @file PaginationMenu component
 * @author Box
 */

import React from 'react';
import range from 'lodash/range';
import Button from 'box-react-ui/lib/components/button';
import DropdownMenu from 'box-react-ui/lib/components/dropdown-menu';
import { Menu, MenuItem } from 'box-react-ui/lib/components/menu';
import { FormattedMessage } from 'react-intl';
import messages from '../messages';
import './PaginationMenu.scss';

type Props = {
    onPageClick: Function,
    pageCount: number,
    pageNumber: number,
};

const PaginationMenu = ({
    onPageClick,
    pageCount = 0,
    pageNumber = 0,
}: Props) => (
    <DropdownMenu
        className="be-pagination-dropdown"
        constrainToWindow
        isRightAligned
    >
        <Button className="be-pagination-toggle">
            <FormattedMessage
                {...messages.pageStatus}
                values={{ pageNumber, pageCount }}
            />
        </Button>
        <Menu className="be-pagination-dropdown-menu">
            {range(1, pageCount + 1).map(page => (
                <MenuItem
                    key={page}
                    isDisabled={page === pageNumber}
                    onClick={() => onPageClick(page)}
                >
                    {page}
                </MenuItem>
            ))}
        </Menu>
    </DropdownMenu>
);

export default PaginationMenu;
