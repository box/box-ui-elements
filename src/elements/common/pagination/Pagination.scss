@import '../variables';

.be-pagination {
    display: flex;

    .be-pagination-count {
        display: flex;
        align-items: center;

        .be-is-small & {
            display: none;
        }
    }

    .be-pagination-toggle {
        margin-right: 5px;
        margin-left: 5px;
    }
}
