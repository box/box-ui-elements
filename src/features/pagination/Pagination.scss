@import '../../styles/variables';

.bdl-Pagination {
    display: flex;

    .bdl-Pagination-count {
        display: flex;
        align-items: center;

        .be-is-small & {
            display: none;
        }
    }

    .bdl-Pagination-toggle {
        margin-right: 5px;
        margin-left: 5px;
    }
}
