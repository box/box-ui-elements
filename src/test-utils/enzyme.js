/* eslint-disable import/no-extraneous-dependencies */
import * as React from 'react';
import { mount as baseMount, shallow as baseShallow } from 'enzyme';

// Global providers that need setup
import { ThemeProvider } from 'styled-components';
import defaultTheme from '../styles/theme';

// Wrap this around every component so they can use app-wide context.
// Prevents breaking due to missing providers.
// See https://github.com/airbnb/enzyme/blob/master/docs/api/ReactWrapper/getWrappingComponent.md
/* eslint-disable react/prop-types */
const Wrappers = ({ children }) => {
    return <ThemeProvider theme={defaultTheme}>{children}</ThemeProvider>;
};

/**
 * mount() from Enzyme but with required app-wide context providers included by default
 */
const mountConnected = (element, options = {}) => baseMount(element, { wrappingComponent: Wrappers, ...options });

/**
 * shallow() from Enzyme but with required app-wide context providers included by default.
 */
const shallowConnected = (element, options = {}) => baseShallow(element, { wrappingComponent: Wrappers, ...options });

// Re-export everything from Enzyme
export * from 'enzyme';
// Export wrapped methods
export { mountConnected, shallowConnected };
