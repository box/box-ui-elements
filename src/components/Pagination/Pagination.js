/**
 * @flow
 * @file Pagination component
 * @author Box
 */

import React from 'react';
import { FormattedMessage } from 'react-intl';
import noop from 'lodash/noop';
import Button from 'box-react-ui/lib/components/button';
import ButtonGroup from 'box-react-ui/lib/components/button-group';
import IconPageBack from 'box-react-ui/lib/icons/general/IconPageBack';
import IconPageForward from 'box-react-ui/lib/icons/general/IconPageForward';
import PaginationMenu from './PaginationMenu';
import Tooltip from '../Tooltip';
import messages from '../messages';
import { DEFAULT_PAGE_SIZE } from '../../constants';
import './Pagination.scss';

type Props = {
    offset?: number,
    onChange?: Function,
    pageSize?: number,
    totalCount?: number,
};

const PAGE_ICON_STYLE = {
    height: 9,
    width: 6,
};

const Pagination = ({
    offset = 0,
    onChange = noop,
    pageSize = DEFAULT_PAGE_SIZE,
    totalCount = 0,
}: Props) => {
    const pageCount = Math.ceil(totalCount / pageSize);
    if (pageCount <= 1) return null;

    const pageByOffset = Math.floor(offset / pageSize) + 1;
    const pageNumber = pageByOffset > 0 ? Math.min(pageCount, pageByOffset) : 1;
    const hasNextPage = pageNumber < pageCount;
    const hasPreviousPage = pageNumber > 1;

    const updateOffset = (newPageNumber: number) => {
        let newOffset = (newPageNumber - 1) * pageSize;

        if (newOffset <= 0) {
            newOffset = 0;
        }

        if (newOffset >= totalCount) {
            newOffset = totalCount - pageSize;
        }

        onChange(newOffset);
    };

    const handleNextClick = () => {
        updateOffset(pageNumber + 1);
    };

    const handlePreviousClick = () => {
        updateOffset(pageNumber - 1);
    };

    return (
        <div className="be-pagination">
            <div className="be-pagination-count">
                <PaginationMenu
                    onPageClick={updateOffset}
                    pageCount={pageCount}
                    pageNumber={pageNumber}
                />
            </div>

            <ButtonGroup>
                <Tooltip
                    isDisabled={!hasPreviousPage}
                    text={<FormattedMessage {...messages.previousPage} />}
                >
                    <Button
                        onClick={handlePreviousClick}
                        isDisabled={!hasPreviousPage}
                    >
                        <IconPageBack {...PAGE_ICON_STYLE} />
                    </Button>
                </Tooltip>
                <Tooltip
                    isDisabled={!hasNextPage}
                    text={<FormattedMessage {...messages.nextPage} />}
                >
                    <Button onClick={handleNextClick} isDisabled={!hasNextPage}>
                        <IconPageForward {...PAGE_ICON_STYLE} />
                    </Button>
                </Tooltip>
            </ButtonGroup>
        </div>
    );
};

export default Pagination;
