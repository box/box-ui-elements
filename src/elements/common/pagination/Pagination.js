/**
 * @flow
 * @file Pagination component
 * @author Box
 */

import React from 'react';
import MarkerBasedPagination from './MarkerBasedPagination';
import OffsetBasedPagination from './OffsetBasedPagination';
import './Pagination.scss';

type Props = {
    hasNextMarker?: boolean,
    hasPrevMarker?: boolean,
    offset?: number,
    onMarkerBasedPageChange?: Function,
    onOffsetChange?: Function,
    pageSize?: number,
    totalCount?: number,
};

const Pagination = ({ hasNextMarker, hasPrevMarker, onMarkerBasedPageChange, ...rest }: Props) => {
    if (hasNextMarker || hasPrevMarker) {
        return (
            <MarkerBasedPagination
                hasNextMarker={hasNextMarker}
                hasPrevMarker={hasPrevMarker}
                onMarkerBasedPageChange={onMarkerBasedPageChange}
            />
        );
    }

    return <OffsetBasedPagination {...rest} />;
};

export default Pagination;
