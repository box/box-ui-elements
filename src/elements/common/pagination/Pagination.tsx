import * as React from 'react';
import MarkerBasedPagination from './MarkerBasedPagination';
import OffsetBasedPagination from './OffsetBasedPagination';
import './Pagination.scss';

export interface PaginationProps {
    hasNextMarker?: boolean;
    hasPrevMarker?: boolean;
    isSmall: boolean;
    offset?: number;
    onMarkerBasedPageChange?: (offset: number) => void;
    onOffsetChange?: (offset: number) => void;
    pageSize?: number;
    totalCount?: number;
}

const Pagination = ({ hasNextMarker, hasPrevMarker, isSmall, onMarkerBasedPageChange, ...rest }: PaginationProps) => {
    if (hasNextMarker || hasPrevMarker) {
        return (
            <MarkerBasedPagination
                hasNextMarker={hasNextMarker}
                hasPrevMarker={hasPrevMarker}
                isSmall={isSmall}
                onMarkerBasedPageChange={onMarkerBasedPageChange}
            />
        );
    }

    return <OffsetBasedPagination isSmall={isSmall} {...rest} />;
};

export default Pagination;
