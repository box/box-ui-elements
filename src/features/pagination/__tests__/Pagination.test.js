import React from 'react';
import { shallow } from 'enzyme';
import Pagination from '../Pagination';

describe('features/pagination/Pagination', () => {
    test.each`
        hasNextMarker | hasPrevMarker
        ${false}      | ${false}
        ${false}      | ${true}
        ${true}       | ${false}
        ${true}       | ${true}
    `(
        'should render correctly depending on the props passed for offsetBased or markerBased pagination',
        ({ hasNextMarker, hasPrevMarker }) => {
            const wrapper = shallow(<Pagination hasNextMarker={hasNextMarker} hasPrevMarker={hasPrevMarker} />);
            expect(wrapper).toMatchSnapshot();
        },
    );

    test('should render OffsetBasedPagination when no markers are provided', () => {
        const wrapper = shallow(<Pagination offset={0} onOffsetChange={() => null} pageSize={10} totalCount={100} />);
        expect(wrapper).toMatchSnapshot();
    });
});
