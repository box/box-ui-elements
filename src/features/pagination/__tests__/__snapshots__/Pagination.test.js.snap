// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`features/pagination/Pagination should render OffsetBasedPagination when no markers are provided 1`] = `
<OffsetBasedPagination
  offset={0}
  onOffsetChange={[Function]}
  pageSize={10}
  totalCount={100}
/>
`;

exports[`features/pagination/Pagination should render correctly depending on the props passed for offsetBased or markerBased pagination 1`] = `<OffsetBasedPagination />`;

exports[`features/pagination/Pagination should render correctly depending on the props passed for offsetBased or markerBased pagination 2`] = `
<MarkerBasedPagination
  hasNextMarker={false}
  hasPrevMarker={true}
/>
`;

exports[`features/pagination/Pagination should render correctly depending on the props passed for offsetBased or markerBased pagination 3`] = `
<MarkerBasedPagination
  hasNextMarker={true}
  hasPrevMarker={false}
/>
`;

exports[`features/pagination/Pagination should render correctly depending on the props passed for offsetBased or markerBased pagination 4`] = `
<MarkerBasedPagination
  hasNextMarker={true}
  hasPrevMarker={true}
/>
`;
